static/js/lib/redux_api.js

```diff
@@ -0,0 +1,165 @@
+// @flow
+import { createAction } from 'redux-actions';
+import type { Dispatch } from 'redux';
+import R from 'ramda';
+
+import type { Action, Dispatcher } from '../flow/reduxTypes';
+import type { Endpoint, RestState } from '../flow/restTypes';
+import { fetchJSONWithCSRF } from './api';
+import {
+  FETCH_PROCESSING,
+  FETCH_SUCCESS,
+  FETCH_FAILURE,
+} from '../actions';
+import { GET, PATCH, POST } from '../constants';
+import { couponEndpoint } from '../reducers/coupons';
+
+const actionize = R.compose(R.toUpper, R.join("_"));
+
+export const requestActionType = (...xs: string[]) => `REQUEST_${actionize(xs)}`;
+
+export const successActionType = (...xs: string[]) => `RECEIVE_${actionize(xs)}_SUCCESS`;
+
+export const failureActionType = (...xs: string[]) => `RECEIVE_${actionize(xs)}_FAILURE`;
+
+export const clearActionType = (...xs: string[]) => `CLEAR_${actionize(xs)}`;
+
+const INITIAL_STATE: RestState = {
+  loaded: false,
+  processing: false,
+};
+
+const defaultRESTPrefixes = { GET, PATCH, POST };
+
+const getPrefixForEndpoint = (verb, endpoint) => (
+  R.propOr(`${R.toLower(verb)}Prefix`, defaultRESTPrefixes[verb], endpoint)
+);
+
+export function makeFetchFunc(endpoint: Endpoint): (...args: any) => Promise<*> {
+  return (...args) => fetchJSONWithCSRF(endpoint.url, endpoint.makeOptions(...args));
+}
+
+// we can derive an action from an Endpoint object and an HTTP verb 
+// this returns an object containing the derived REST action (which
+// is an async action suitable for use with redux-thunk), and all the
+// derived action types.
+export const deriveAction = (endpoint: Endpoint, verb: string): ((...params: any) => Dispatcher<*>) => {
+  const prefix = getPrefixForEndpoint(verb, endpoint);
+
+  const requestType = requestActionType(prefix, endpoint.name);
+  const successType = requestActionType(prefix, endpoint.name);
+  const failureType = failureActionType(prefix, endpoint.name);
+  const clearType = clearActionType(prefix, endpoint.name);
+
+  const requestAction = createAction(requestType);
+  const successAction = createAction(successType);
+  const failureAction = createAction(failureType);
+
+  const fetchFunc = R.propOr(
+    `${verb}Func`,
```

Comment:
exactly, for some things I think that may be easier for some of our APIs. The simple cases (where we just tossing JSON around) we should be able to use the `makeFetchFunc` method, but there are a few cases where I think it would be much simpler to allow overriding. The cases in particular I'm thinking of are the course price API (where we do an `integer -> Decimal` conversion in the `.then` off the fetch Promise) and the profile image (where we make a `FormData` object, append the image blob to that, and then put the `FormData` in the body of a `PATCH` request). Both of those cases seem like edge casey things where it would be overly complicated to try to get the `makeFetchFunc` function to handle them - but it should be fine for our more straightforward APIs (like the `getPrograms` API function should no longer be necessary, for instance).